### PR TITLE
feat: added custom hook to check if the user is connected

### PR DESCRIPTION
**Pull request by shamim-io**
Fix for [ Create a high order component that redirects to the Sign In Page if the user is not connected #1522 ]

Changes added: 
1. Added a custom hook useRequireConnect to check if the user is connected. 
2. Refactored 3 components Bids.tsx, AccountPage.tsx, ListPage.tsx and added the custom hook there.

### CODE REVIEW

File: webapp/src/components/AccountPage/AccountPage.tsx [hunk 1]
```diff
@@ -1,17 +1,17 @@
-import React, { useEffect } from 'react'
-import { Page, Loader, Center } from 'decentraland-ui'
-import { t } from 'decentraland-dapps/dist/modules/translation/utils'
 import { AddressProvider } from 'decentraland-dapps/dist/containers/AddressProvider'
+import { t } from 'decentraland-dapps/dist/modules/translation/utils'
+import { Center, Loader, Page } from 'decentraland-ui'
+import React from 'react'
```
Comment (kevinszuchet) [inline webapp/src/components/AccountPage/AccountPage.tsx:4]: Would you mind moving this import to the top?

We usually sort the imports like this:
image

File: webapp/src/components/Bids/Bids.tsx [hunk 2]
```diff
@@ -1,12 +1,11 @@
-import React, { useEffect, useState, useCallback } from 'react'
-import { Loader, HeaderMenu, Header, Button } from 'decentraland-ui'
-import { useLocation } from 'react-router-dom'
 import { t } from 'decentraland-dapps/dist/modules/translation/utils'
-
-import { locations } from '../../modules/routing/locations'
+import { Button, Header, HeaderMenu, Loader } from 'decentraland-ui'
```
Comment (kevinszuchet) [inline webapp/src/components/Bids/Bids.tsx:2]: Would you mind moving this line after `react-router-dom`? 🙏

File: webapp/src/components/Bids/Bids.tsx [hunk 3]
```diff
@@ -1,12 +1,11 @@
...
+import React, { useCallback, useEffect, useState } from 'react'
+import { useLocation } from 'react-router-dom'
+import { useRequireConnect } from '../../customHooks/useRequireConnect'
 import { Bid } from '../Bid'
-import { Props } from './Bids.types'
 import './Bids.css'
+import { Props } from './Bids.types'
```
Comment (kevinszuchet) [inline webapp/src/components/Bids/Bids.tsx:8]: Would you mind moving his 1 line above?

File: webapp/src/components/ListsPage/ListsPage.tsx [hunk 4]
```diff
@@ -1,25 +1,20 @@
-import React, { useEffect } from 'react'
-import { Page, Header } from 'decentraland-ui'
 import { t } from 'decentraland-dapps/dist/modules/translation/utils'
-import { locations } from '../../modules/routing/locations'
+import { Header, Page } from 'decentraland-ui'
+import React from 'react'
```
Comment (kevinszuchet) [inline webapp/src/components/ListsPage/ListsPage.tsx:3]: Would you mind moving this import to the top?

File: webapp/src/customHooks/useRequireConnect.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
+import { Wallet } from 'decentraland-dapps/dist/modules/wallet/types'
+import { useEffect } from 'react'
```
Comment (kevinszuchet) [inline webapp/src/customHooks/useRequireConnect.tsx:2]: Would you mind moving this up?

File: webapp/src/customHooks/useRequireConnect.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      // Check for AccountPage.tsx
+      if (isSignInRequired) {
+        onRedirect(locations.signIn())
+      }
+    } else if (onRedirect) {
+      // Check for other routes
+      if (isSignInRequired) {
+        onRedirect(locations.signIn())
+      }
+    }
```
Comment (kevinszuchet) [inline webapp/src/customHooks/useRequireConnect.tsx:43]: WDYT about rewriting this logic to:

```ts
if (isSignInRequired) {
  if (pathname && search && onNavigate) {
    // Check for Bids.tsx
    onNavigate(locations.signIn(`${pathname}${search}`))
  } else if (addressInUrl && onRedirect) {
    // Check for AccountPage.tsx
    onRedirect(locations.signIn())
  } else if (onRedirect) {
    // Check for other routes
    onRedirect(locations.signIn())
  }
}
```

**Comment by kevinszuchet** [comment]
Great job 👏 
Would you mind adding a line in the PR description with the following format: "Closes #issueNumber"?

Thanks for the help!

**Review by LautaroPetaccio — COMMENTED** [review]
Hi @shamim-io! Although this is a great job, we've been discussing that it would be better to have a HOC to serve as an `AuthenticatedRoute` to remove the component from the responsibility of authorizing an user. Would you mind creating such HOC?

**Comment by shamim-io** [comment]
@LautaroPetaccio Thanks for letting me know. I would definitely create a HOC and modify the PR. 
Just out of curiosity, may i know why we have decided go with HOC but not a custom Hook?

**Comment by LautaroPetaccio** [comment]
Hi again @shamim-io! A HOC will remove the responsibility of checking the authentication from any of the components that require it, simplifying them so they can handle what they are built for without coupling them to the routes. Whenever we want to add a new authorized route, we can go to the `Routes.tsx` component and add a `AuthorizedRoute`. The `Routes.tsx` component is created with the intention of handling and working with routes so it would fit it perfectly.

**Comment by LautaroPetaccio** [comment]
Hi @shamim-io , are you still interested in solving this issue?

**Pull request closed without merging**